File: T-Bank/Algorithms-and-Data-Structures-2025/7sprint-DynamicProgramming/contest/A-Платная_лестница/solution.cpp
#include <iostream>
#include <vector>


void solution() {
    /*
    Time Complexity O(N)
    Memory Complexity O(N)
    */
    int n;
    std::scanf("%d", &n);
    std::vector<long long> nums (n + 1, 0);
    for (int i = 1; i <= n; ++i) std::scanf("%lld", &nums[i]);
    std::vector<long long> dp (n + 1, 0);
    dp[1] = nums[1];
    for (int level = 2; level <= n; ++level) {
        dp[level] = std::min(dp[level - 1], dp[level - 2]) + nums[level];
    }
    std::printf("%lld\n", dp[n]);
}


int main() {
    solution();
}

File: T-Bank/Algorithms-and-Data-Structures-2025/7sprint-DynamicProgramming/contest/A-Платная_лестница/solution.py
import sys


def solution() -> None:
    """
    Time Complexity O(N)
    Memory Complexity O(N)
    """
    n: int = int(sys.stdin.readline().rstrip())
    nums: list[int] = [0] + list(map(int, sys.stdin.readline().rstrip().split()))
    dp: list[int] = [0] * (n + 1)
    dp[1] = nums[1]
    for level in range(2, n + 1):
        dp[level] = min(dp[level - 1], dp[level - 2]) + nums[level]
    sys.stdout.write('{}\n'.format(dp[n]))


if __name__ == '__main__':
    solution()